lib/common/rules.c

```diff
@@ -1155,41 +1155,77 @@ pcmk__evaluate_attr_expression(const xmlNode *expression,
  * \internal
  * \brief Evaluate an operation rule expression
  *
- * \param[in] expr        XML of a rule's \c PCMK_XE_OP_EXPRESSION subelement
- * \param[in] rule_input  Values used to evaluate rule criteria
+ * \param[in] op_expression  XML of a rule's \c PCMK_XE_OP_EXPRESSION subelement
+ * \param[in] rule_input     Values used to evaluate rule criteria
  *
- * \return Standard Pacemaker return code
+ * \return Standard Pacemaker return code (\c pcmk_rc_ok if the expression
+ *         passes, some other value if it does not)
  */
 int
-pcmk__evaluate_op_expression(const xmlNode *expr,
+pcmk__evaluate_op_expression(const xmlNode *op_expression,
                              const pcmk_rule_input_t *rule_input)
 {
-    const char *name = crm_element_value(expr, PCMK_XA_NAME);
-    const char *interval_s = crm_element_value(expr, PCMK_META_INTERVAL);
+    const char *id = NULL;
+    const char *name = NULL;
+    const char *interval_s = NULL;
     guint interval_ms = 0U;
 
-    crm_trace("Testing op_defaults expression: %s", pcmk__xe_id(expr));
+    if ((op_expression == NULL) || (rule_input == NULL)) {
+        return EINVAL;
+    }
 
-    if (rule_input->op_name == NULL) {
-        crm_trace("No operation data provided");
-        return pcmk_rc_op_unsatisfied;
+    // Get operation expression ID (for logging)
+    id = pcmk__xe_id(op_expression);
+    if (pcmk__str_empty(id)) { // Not possible with schema validation enabled
+        /* @COMPAT When we can break behavioral backward compatibility,
+         * return pcmk_rc_op_unsatisfied
+         */
+        pcmk__config_warn(PCMK_XE_OP_EXPRESSION " element has no " PCMK_XA_ID);
+        id = "without ID"; // for logging
+    }
+
+    // Validate operation name
+    name = crm_element_value(op_expression, PCMK_XA_NAME);
+    if (pcmk__str_empty(name)) { // Not possible with schema validation enabled
```

Comment:
Research yielded some surprises.

An empty string here does pass schema validation. I expected that a required attribute must be non-empty, but RNG's `<text/>` (the default content for attributes) accepts empty strings. It should be possible to require non-empty strings or even matching a regular expression, using `<data type="string">` with extra parameters, but libxml2 appears to ignore parameters in a quick test.

Some RNG types (such as ID and `<value>`) normalize whitespace (stripping leading and trailing whitespace, and changing whitespace sequences to a single space) before validating. None of our code considers that. I'm not sure whether we should do anything about it now, but it does mean configurations can pass validation that will be rejected by the code.

Even before validation, XML parsers are supposed to normalize attribute values (https://www.w3.org/TR/xml11/#AVNormalize). libxml2 supposedly follows this, so I'm not sure why we don't see that.

Since empty strings are allowed here by RNG, I'm leaning to just checking NULL and letting the comparison fail.

Reply:
> Since empty strings are allowed here by RNG, I'm leaning to just checking NULL and letting the comparison fail.

Yeah I think that's what I would do. I don't care much, I just want to make you aware of any changes or edge cases I find so that you can make an informed decision.

And I know we can't make the schema tighter at-will.

Reply:
I tried something like:

    <data type="string">
       <param name="minLength">1</param>
    </data>

which works with xmllint but appeared to be ignored when I tried it in the CIB schema (I didn't put a lot of time into it, so I may have messed it up). The `except` method doesn't rely on XSD datatypes so that may be better for that particular use case.

> Steps 1 through 3 of AV normalization apply,

Then why did we ever have to worry about newline escaping? Why didn't libxml2 provide us with already-escaped content?

Reply:
Ah okay. I've *only* tried with `xmllint` so far. Yeah `except` might work, no need to figure it out right now though.

---

> Why didn't libxml2 provide us with already-escaped content?

As best I remember[*] libxml2 simply doesn't do that for newlines. When serializing XML, it escapes characters that would make the output invalid, such as `'<'` and `'&'`, but not newlines.

[*] I use this qualifier because there are two libxml2 public API functions that are supposed to escape special characters, and IIRC there are more places internal to libxml2 that perform their own escaping... some of the behavior is questionable and it's hard to keep track of.

To further complicate things, we don't use libxml2 to serialize the whole XML document. In `pcmk__xml_string()` and its helpers, we dump the document piece-by-piece because we often need to filter out certain things, and libxml2 doesn't have that capability.

Concretely:
1. Set an attribute to `"some\nval"` (with literal newline). It's stored in memory as an `xmlAttr` with the literal newline.
2. Dump the XML document to text. Now there's a line break in the output:
```
test_attr="some
val"
```
3. Parse the XML document. Now the literal newline becomes a space. To prevent that from happening, the dumped text needs to have `test_attr="some&#x0A;val"`.

For a long time, we escaped all newlines as backslash-n when dumping them, so step 2's output had `test_attr="some\nval"`. Every future parse maintained and stored the literal backslash-n, and every future dump outputted it (without further escaping).

When we stopped escaping as backslash-n, we got the behavior demonstrated in the steps above.

Note that sending an IPC request dumps the request XML to text, and the IPC server parses it. So a sufficient condition for losing a newline is sending the XML via IPC. (There are other scenarios that involve dumping and parsing but that's a very common one.)

Reply:
> we don't use libxml2 to serialize the whole XML document

For a while we did, but only if `pcmk__xml_fmt_text` was set. That resulted in some output inconsistencies and prevented us from using const, so we dropped it.

There is one place (`pcmk__xml2fd()`) where we still use libxml2 to serialize the whole document. I've argued against keeping that before. We've dropped all but one caller (in `output_xml.c:xml_finish()`).
* I don't want to maintain it (though it is small)
* I worry about any inconsistencies that may arise compared to our usual custom dumping
* Any efficiency gain doesn't make a difference in `out->finish()`

Reply:
Looking back, you were replying to "Steps 1 through 3 of AV normalization apply", so I might not have answered your question. Just in case:

If we're parsing a literal newline, we hit this:
> For a white space character (#x20, #xD, #xA, #x9), append a space character (#x20) to the normalized value.

If we're parsing `"&#x0A;"` (or equivalent), we hit this:
> For a character reference, append the referenced character to the normalized value.

This section doesn't talk about escaping but rather about how characters are parsed -- which differs depending on whether they were *already* escaped as character references or not

Reply:
Right, so I would expect that libxml2 would replace newlines with spaces when the daemon parses the API request to set a multiline node attribute. But from what I gather from your previous comments, libxml2 just doesn't comply with newline normalization.

Reply:
> But from what I gather from your previous comments, libxml2 just doesn't comply with newline normalization.

It does. The point I'm making is that parsing XML from text is different from using `xmlSetProp()` or `crm_xml_add()` to set an attribute directly.

It normalizes newlines just fine when it's parsing them from text. If it's a literal newline, it converts it to a space. If it's a character reference, it preserves the newline. Demo (added to `crm_error.c` as a hack):
```
    {
        const char *s1 = "<n attr=\"a\nb\"/>";
        const char *s2 = "<n attr=\"a&#x0A;b\"/>";

        xmlNode *node = pcmk__xml_parse(s1);
        xmlAttr *attr = pcmk__xe_first_attr(node);

        for (const char *v = pcmk__xml_attr_value(attr); *v != '\0'; v++) {
            fprintf(stderr, "%c", *v);
        }
        fprintf(stderr, "\n");
        free_xml(node);

        node = pcmk__xml_parse(s2);
        attr = pcmk__xe_first_attr(node);

        for (const char *v = pcmk__xml_attr_value(attr); *v != '\0'; v++) {
            fprintf(stderr, "%c", *v);
        }
        fprintf(stderr, "\n");
        free_xml(node);
    }
```

Output:
```
$ crm_error asdf
a b
a
b
```

Note that the literal newline got parsed to a space, and the character reference was preserved as a newline.

---

By contrast, if I set the attribute directly (instead of parsing it from text), the literal newline is preserved. No need for a character reference. In fact, the character reference gets treated as a literal string instead of as a newline.
```
    {
        xmlNode *node = pcmk__xe_create(NULL, "n");
        xmlAttr *attr = NULL;

        xmlSetProp(node, (pcmkXmlStr) "attr", (pcmkXmlStr) "a\nb");
        attr = pcmk__xe_first_attr(node);

        for (const char *v = pcmk__xml_attr_value(attr); *v != '\0'; v++) {
            fprintf(stderr, "%c", *v);
        }
        fprintf(stderr, "\n");

        xmlSetProp(node, (pcmkXmlStr) "attr", (pcmkXmlStr) "a&#x0A;b");
        attr = pcmk__xe_first_attr(node);

        for (const char *v = pcmk__xml_attr_value(attr); *v != '\0'; v++) {
            fprintf(stderr, "%c", *v);
        }
        fprintf(stderr, "\n");

        free_xml(node);
    }
```

Output:
```
$ crm_error asdf
a
b
a&#x0A;b
```

Reply:
(ignore the weird "for" loop -- i initially had a more complicated printing scheme and then realized I didn't need it)

Reply:
> > But from what I gather from your previous comments, libxml2 just doesn't comply with newline normalization.
> 
> It does. The point I'm making is that parsing XML from text is different from using `xmlSetProp()` or `crm_xml_add()` to set an attribute directly.

Right, so why doesn't pacemaker-attrd normalize newlines when the attribute is parsed from the IPC XML, set the normalized value in its memory, and return that when queried?

Reply:
> Right, so why doesn't pacemaker-attrd normalize newlines when the attribute is parsed from the IPC XML, set the normalized value in its memory, and return that when queried?

It does. Before #3395 (note the newline is normalized as documented):
```
reid@laptop:~/git/pacemaker$ sudo attrd_updater -N laptop -n test_attr -U "$(echo -e "  a\nb  ")"
reid@laptop:~/git/pacemaker$ sudo attrd_updater -N laptop -n test_attr -Q
name="test_attr" host="laptop" value="  a b  "
```